### PR TITLE
 Doc support for salted SHA-256 algorithm for password import 

### PR DESCRIPTION
**Pull request by mpeng-okta**
## Description:
- Doc support for salted SHA-256 algorithm for password import 

### Resolves:

* [OKTA-158149](https://oktainc.atlassian.net/browse/OKTA-158149)

## Screenshots:
image
image

## Reviewers:
- @mystiberry-okta 
- @srinivasanagandla-okta 
- @edjohns-okta

### CODE REVIEW

**Comment by mystiberry-okta** [comment]
@mpeng-okta  Can we merge this for 2018.09? Wait for Srini or 2018.10?

**Comment by mpeng-okta** [comment]
@mystiberry-okta ah yeah, wait for Srini as well as 2018.10 (the code change isn't in just yet, and I'm aiming to get that in for 2018.10). Thanks!

File: _source/_docs/api/resources/users.md [hunk 1]
```diff
@@ -3799,13 +3799,15 @@ The password specified in the value property must meet the default password poli
...
-| workFactor    | Integer   | Governs the strength of the hash, and the time required to compute it | FALSE     | 1  | 20  |
-| salt    | String   | Specifies the password salt used to generate the hash | FALSE     | 22  | 22  |
-| value | String | The actual hashed password| FALSE | N/A | N/A |
+| Property   | DataType | Description                                                                                                 | Required                           | Min Value          | Max Value          |
+|:-----------|:---------|:------------------------------------------------------------------------------------------------------------|:-----------------------------------|:-------------------|:-------------------|
+| algorithm  | String   | The algorithm used to hash the password. Must be set to `BCRYPT` or `SHA-256`                               | TRUE                               | N/A                | N/A                |
+| value      | String   | The actual hashed password                                                                                  | TRUE                               | N/A                | N/A                |
```
Comment (srinivasanagandla-okta) [inline _source/_docs/api/resources/users.md]: Is the value just hashed or hashed and base64 encoded?

File: _source/_docs/api/resources/users.md [hunk 2]
```diff
@@ -3818,10 +3820,24 @@ A hashed password may be specified in a Password Object when creating or updatin
...
+    "salt": "MPu13OmY",
+    "saltOrder": "PREFIX",
+    "value": "Gjxo7mxvvzQWa83ovhYRUH2dWUhC1N77Ntc56UfI4sY"
+  }
+}
+~~~
 
 ##### Hashing Function
 
-Okta supports the bcrypt hashing function.
```
Comment (srinivasanagandla-okta) [inline _source/_docs/api/resources/users.md]: I am not sure if we need this section. We seem to be revealing details that are not required here.

File: _source/_docs/api/resources/users.md [hunk 3]
```diff
@@ -3799,13 +3799,15 @@ The password specified in the value property must meet the default password poli
...
+| algorithm  | String   | The algorithm used to hash the password. Must be set to `BCRYPT` or `SHA-256`                               | TRUE                               | N/A                | N/A                |
+| value      | String   | The actual hashed password                                                                                  | TRUE                               | N/A                | N/A                |
+| salt       | String   | Specifies the password salt used to generate the hash                                                       | TRUE                               | 22 (only for `BCRYPT` algorithm) | 22 (only for `BCRYPT` algorithm) |
+| workFactor | Integer  | Governs the strength of the hash, and the time required to compute it. Only relevant for `BCRYPT` algorithm | Only for `BCRYPT` algorithm        | 1                  | 20                 |
+| saltOrder  | String   | Specifies whether salt was pre- or postfixed to password before hashing. Only relevant for `SHA-256` algorithm. Must be set to `PREFIX` or `POSTFIX` | Only for `SHA-256` algorithm      | N/A                | N/A                |
 
+###### BCRYPT-Hashed Password Object Example
```
Comment (srinivasanagandla-okta) [inline _source/_docs/api/resources/users.md]: Do we need the `-` between `BCRYPT` and `Hashed`?

File: _source/_docs/api/resources/users.md [hunk 4]
```diff
@@ -3799,13 +3799,15 @@ The password specified in the value property must meet the default password poli
...
-| value | String | The actual hashed password| FALSE | N/A | N/A |
+| Property   | DataType | Description                                                                                                 | Required                           | Min Value          | Max Value          |
+|:-----------|:---------|:------------------------------------------------------------------------------------------------------------|:-----------------------------------|:-------------------|:-------------------|
+| algorithm  | String   | The algorithm used to hash the password. Must be set to `BCRYPT` or `SHA-256`                               | TRUE                               | N/A                | N/A                |
+| value      | String   | The actual hashed password                                                                                  | TRUE                               | N/A                | N/A                |
+| salt       | String   | Specifies the password salt used to generate the hash                                                       | TRUE                               | 22 (only for `BCRYPT` algorithm) | 22 (only for `BCRYPT` algorithm) |
```
Comment (srinivasanagandla-okta) [inline _source/_docs/api/resources/users.md]: What is the min, max value referring to? This doesn't seem right.

File: _source/_docs/api/resources/users.md (hunk 1 above)
Comment (mpeng-okta) [reply _source/_docs/api/resources/users.md]: Looking in okta-core, it appears that the value is hashed and base64 encoded. It just originally said "The actual hashed password", so I left it as such. Shall I update the description

File: _source/_docs/api/resources/users.md (hunk 2 above)
Comment (mpeng-okta) [reply _source/_docs/api/resources/users.md]: Do you think we should just remove the entirety of the "Hashing Function" section?

File: _source/_docs/api/resources/users.md (hunk 3 above)
Comment (mpeng-okta) [reply _source/_docs/api/resources/users.md]: I don't think so. I can remove it.

File: _source/_docs/api/resources/users.md (hunk 4 above)
Comment (mpeng-okta) [reply _source/_docs/api/resources/users.md]: From the existing documentation, it looks like it's saying the BCRYPT salt length has to be exactly 22 characters. The example seems to be consistent with this description.

However, I agree that it doesn't seem right. I did some digging, and it appears that the max length for the salt is 22, to support the full 128-bit salt space (https://www.usenix.org/legacy/events/usenix99/provos/provos_html/node6.html#SECTION00051000000000000000). I'm still not sure exactly about salt length _requirements_, and looking at BCrypt.class, I still wasn't able to elucidate required salt length.

File: _source/_docs/api/resources/users.md (hunk 2 above)
Comment (mystiberry-okta) [reply _source/_docs/api/resources/users.md]: We were asked to clarify this for customers who wanted to know what hashing algo/function we supported.

Comment (srinivasanagandla-okta) [reply _source/_docs/api/resources/users.md]: @mpeng-okta I agree we can talk about algorithms that we support via API but I am not sure if we need to talk about how we internally store it.

**Comment by srinivasanagandla-okta** [comment]
@edjohns-okta Can you also take a look?

File: _source/_docs/api/resources/users.md (hunk 2 above)
Comment (mpeng-okta) [reply _source/_docs/api/resources/users.md]: Sounds good to me.

Comment (mpeng-okta) [reply _source/_docs/api/resources/users.md]: This has been updated accordingly, though the old diff still shows up since I'm still removing this line.

**Comment by mystiberry-okta** [comment]
@mpeng-okta shall I merge? From weekly branch, things are published a few hours before 2018.10 goes out.

**Comment by mpeng-okta** [comment]
@mystiberry-okta yes, please do!

**Pull request merged**